src/com/unbounce/treajure/concurrent/mvar.clj

```diff
@@ -0,0 +1,82 @@
+(ns com.unbounce.treajure.concurrent.mvar
+  "Haskell's MVar abstraction in Clojure.
+
+  Check: http://chimera.labs.oreilly.com/books/1230000000929/ch07.html for more
+  info around MVars."
+  (:import [java.util.concurrent Semaphore]))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Lightweight Performance mutable tuple implementation in Clojure
+;; (Look ma! no atoms!)
+
+(defprotocol IMCons
+  (mcar [p])
+  (mcdr [p])
+  (set-mcar! [p val])
+  (set-mcdr! [p val]))
+
+(deftype MCons [^{:volatile-mutable true} car
+                ^{:volatile-mutable true} cdr]
```

Comment:
Those are semantics from lisp,  car means head,  cdr means tail,  I didn't write this bit myself,  found it out in a gist from guy that implemented this from emulating the racket library. 

After talking with Arthur,  we probably are going to replace this implementation with a simpler blocking queue of size 1, discovered this was an option after this was pushed